File: 1881E.py
 
def rec(el,n,dict = {}):
    if(n==0 or n==1):
        return 0
    if(tuple(el)in dict):
        return dict[tuple(el)]
    if(el[0]<=n-1):
        dict[tuple(el)] = min(1+rec(el[1:],n-1,dict),rec(el[el[0]+1:],n-el[0]-1,dict))
        return min(1+rec(el[1:],n-1,dict),rec(el[el[0]+1:],n-el[0]-1,dict))
    else:
        dict[tuple(el)] = 1+rec(el[1:],n-1,dict)
        return 1+rec(el[1:],n-1,dict)
    




for i in range(int(input())):
    n = int(input())
    el = list(map(int, input().split()))
    print(rec(el,n))

    
    

File: 1881E.cpp
#include <bits/stdc++.h>

using namespace std;
int rec(vector<int>el,int n,map<vector<int>,int> a={}){
    if(n==0){
        return 0;
    }
    if(a.find(el)!=a.end()){
        return a[el];
    }
    
    if(el[0]<=n-1){
        vector <int> el1 (el.begin()+1,el.end());
        vector <int> el2 (el.begin()+el[0]+1,el.end());
        a[el] = min(1+rec(el1,n-1,a),rec(el2,n-el[0]-1,a));
        return min(1+rec(el1,n-1,a),rec(el2,n-el[0]-1,a));
    }
    else{
        vector <int> el1 (el.begin()+1,el.end());
        
        a[el] = 1+rec(el1,n-1,a);
        return 1+rec(el1,n-1,a);
    }
}   


int main(){
    int n;
    int t;
    cin>>t;
    while(t--){
        cin >> n;
        vector<int> a(n);
        for(int i =0;i<n;i++){
            cin>>a[i];
        }

        cout << rec(a,n) << '\n';
    }
    


    return 0;
}



